src/native/libs/System.Security.Cryptography.Native/pal_evp_pkey.c

```diff
@@ -275,3 +275,40 @@ int32_t CryptoNative_EncodeSubjectPublicKeyInfo(EVP_PKEY* pkey, uint8_t* buf)
     ERR_clear_error();
     return i2d_PUBKEY(pkey, &buf);
 }
+
+static EVP_PKEY* LoadKeyFromEngine(
+    const char* engineName,
+    const char* keyName,
+    ENGINE_LOAD_KEY_PTR load_func)
+{
+    ERR_clear_error();
+
+    EVP_PKEY* ret = NULL;
+    ENGINE* engine = NULL;
+
+    engine = ENGINE_by_id(engineName);
+
+    if (engine != NULL)
+    {
+        if (ENGINE_init(engine))
```

Comment:
The structural reference of the `engine` is supposed to also be taken internally when loading the object, it might be a good idea to never release it until the key stops being used.

I'm not sure the engine initialization and deinitialization is correct: my interpretation of https://www.openssl.org/docs/man1.0.2/man3/engine.html is that you must maintain the engine initialized for the duration of the key handle being in-use. (E.g. if TLS is using the handle, the functional reference should still be available - i.e. if this key is used in conjunction with `SslStream`.)

Reply:
Per https://github.com/openssl/openssl/discussions/21427 this is safe

Reply:
Thanks for getting the confirmation on that. I think this would be useful to capture as a code comment (perhaps with a pointer back to the openssl discussion you asked).